File: cpp/file_whisper_lib/analyzer.cc
#include "analyzer.hpp"

namespace analyzer
{
    using namespace whisper_data_type;

    void analyze_compressed_file(std::shared_ptr<Node> node)
    {
        std::vector<uint8_t> data;
        if (std::holds_alternative<File>(node->content))
        {
            File &file = std::get<File>(node->content);
            data = file.content;
        }
        else if (std::holds_alternative<Data>(node->content))
        {
            spdlog::debug("extract_compressed_file enter Data type");
        }

        std::locale::global(std::locale("en_US.UTF-8"));
        try
        { // bit7z classes can throw BitException objects
            using namespace bit7z;

            const char *path = std::getenv("LIB_PATH_7Z");
            if (path == NULL)
            {
                path = "/usr/local/share/7z/7z.so";
            }
            Bit7zLibrary lib{path};

            BitArchiveReader arc{lib, data, BitFormat::Auto};

            node->meta.map_number["items_count"] = arc.itemsCount();
            node->meta.map_number["folders_count"] = arc.foldersCount();
            node->meta.map_number["files_count"] = arc.filesCount();
            node->meta.map_number["size"] = arc.size();
            node->meta.map_number["pack_size"] = arc.packSize();
            node->meta.map_bool["is_encrypted"] = arc.isEncrypted();
            node->meta.map_number["volumes_count"] = arc.volumesCount();
            node->meta.map_bool["is_multi_volume"] = arc.isMultiVolume();
        }
        catch (const bit7z::BitException &ex)
        { /* Do something with ex.what()...*/
            spdlog::error(ex.what());
            // std::cerr << "Unknown bit7z error: " << ex.what() << std::endl;
            throw;
        }
    }

}

File: src/file_whisper_lib/analyzer.py
import pybit7z
import logging
import tempfile
from typing import List, Dict
from .dt import Node, File, Data

class Analyzer:
    @staticmethod
    def analyze_compressed_file(node: Node):
        # data_stream = BytesIO()
        # if isinstance(node.content, File):
        #     data_stream = BytesIO(node.content.content)
        # elif isinstance(node.content, Data):
        #     logging.debug("analyze_compressed_file enter Data type")
        #     return
            
        # with pybit7z.lib7zip_context() as lib:
        #     arc = pybit7z.BitArchiveReader(lib, data_stream.getvalue(), pybit7z.FormatAuto)
        #     node.meta.map_number["items_count"] = arc.items_count()
        #     node.meta.map_number["folders_count"] = arc.folders_count()
        #     node.meta.map_number["files_count"] = arc.files_count()
        #     node.meta.map_number["size"] = arc.size()
        #     node.meta.map_number["pack_size"] = arc.pack_size()
        #     node.meta.map_bool["is_encrypted"] = arc.is_encrypted()
        #     node.meta.map_number["volumes_count"] = arc.volumes_count()
        #     node.meta.map_bool["is_multi_volume"] = arc.is_multi_volume()
    #         @typing.overload
    # def __init__(
    #     self,
    #     library: Bit7zLibrary,
    #     in_archive: bytes,
    #     format: BitInFormat = ...,
    #     password: str = "",
    # ) -> None: 不生效，只能生成临时文件来读取

        if isinstance(node.content, File):
            if not node.content.content:
                logging.error("Empty content")
                return
            node.meta.map_bool["is_encrypted"] = True
                
            try:
                with tempfile.NamedTemporaryFile(delete=True) as temp_file:
                    temp_file.write(node.content.content)
                    temp_file.flush()
                    
                    with pybit7z.lib7zip_context() as lib:
                        print(temp_file.name)
                        arc = pybit7z.BitArchiveReader(lib, temp_file.name, pybit7z.FormatAuto)
                        node.meta.map_number["items_count"] = arc.items_count()
                        node.meta.map_number["folders_count"] = arc.folders_count()
                        node.meta.map_number["files_count"] = arc.files_count() 
                        node.meta.map_number["size"] = arc.size()
                        node.meta.map_number["pack_size"] = arc.pack_size()
                        node.meta.map_bool["is_encrypted"] = arc.is_encrypted()
                        node.meta.map_number["volumes_count"] = arc.volumes_count()
                        node.meta.map_bool["is_multi_volume"] = arc.is_multi_volume()
                        
            except pybit7z.BitException as e:
                logging.error(f"Failed to analyze compressed file: {str(e)}")
                return
                
        elif isinstance(node.content, Data):
            logging.debug("analyze_compressed_file enter Data type")
            return
